broker.go

```diff
@@ -0,0 +1,584 @@
+	// cancelBroker cancels broker goroutines
+	cancelBrkr context.CancelFunc
+	// cancelAll cancels pubsub and datatransfer and broker goroutines
+	cancelAll context.CancelFunc
+}
+
+// ChangeEvent notifies an OnChange reader that a specified peer completed a sync.
+type ChangeEvent struct {
+	Cid    cid.Cid
+	PeerID peer.ID
+}
+
+// Defined tells whether the ChangeEvent holds any data.
+func (ce ChangeEvent) Defined() bool {
+	return ce.Cid.Defined()
+}
+
+// handler holds state that is specific to a peer
+type handler struct {
+	dtManager dt.Manager
+	// distEvents is used to communicate a syncDone event back to the
+	// LegBroker for distribution to OnChange readers.
+	distEvents chan<- ChangeEvent
+	latestSync ipld.Link
+	msgChan    chan cid.Cid
+	// peerID is the ID of the peer this handler is responsible for.
+	peerID peer.ID
+	// putSyncDone is a function supplied by the LegBroker for the handler
+	// to give LegBroker the channel to send sync done notification to.
+	putSyncDone func(cid.Cid, chan<- struct{})
+	// syncMutex serializes the handling of individual syncs
+	syncMutex sync.Mutex
+}
+
+// NewLegBroker creates a new LegBroker, and will process messages that are allowed by given peerPolicy.
+func NewLegBroker(host host.Host, ds datastore.Batching, lsys ipld.LinkSystem, topic string, dss ipld.Node, addrTTL time.Duration, peerPolicy PeerPolicyHandler) (*LegBroker, error) {
+	ctx, cancelAll := context.WithCancel(context.Background())
+
+	pubsubTopic, err := makePubsub(ctx, host, topic)
+	if err != nil {
+		cancelAll()
+		return nil, err
+	}
+
+	psub, err := pubsubTopic.Subscribe()
+	if err != nil {
+		cancelAll()
+		return nil, err
+	}
+
+	dt, gs, tmpDir, err := makeDataTransfer(ctx, host, ds, lsys)
+	if err != nil {
+		cancelAll()
+		return nil, err
+	}
+
+	if addrTTL == 0 {
+		addrTTL = defaultAddrTTL
+	}
+
+	ctx, cancelBrkr := context.WithCancel(ctx)
+
+	lb := &LegBroker{
+		dtManager:  dt,
+		host:       host,
+		gsExchange: gs,
+		tmpDir:     tmpDir,
+
+		addrTTL:    addrTTL,
+		psub:       psub,
+		topic:      pubsubTopic,
+		cancelBrkr: cancelBrkr,
+		cancelAll:  cancelAll,
+
+		peerPolicy: peerPolicy,
+		handlers:   make(map[peer.ID]*handler),
+		inEvents:   make(chan ChangeEvent, 1),
+	}
+
+	lb.unsubEvents = dt.SubscribeToEvents(lb.onEvent)
+
+	// Start watcher to read pubsub messages.
+	go lb.watch(ctx)
+	// Start distributor to send ChangeEvent messages to interested parties.
+	go lb.distributeEvents(ctx)
+
+	return lb, nil
+}
+
+// GetLatestSync returns the latest synced CID for the specified peer. If there
+// is not handler for the peer, then nil is returned.  This does not mean that
+// no data is synced with that peer, it means that the broker does not know
+// about it.  Calling Sync() first may be necessary.
+func (lb *LegBroker) GetLatestSync(peerID peer.ID) ipld.Link {
+	lb.handlersMutex.Lock()
+	defer lb.handlersMutex.Unlock()
+
+	hnd, ok := lb.handlers[peerID]
+	if !ok {
+		return nil
+	}
+
+	return hnd.latestSync
```

Comment:
Could there be a race condition for reading this field via `GetLatestSync` at broker level and at handler level?

Reply:
Fixed